ssl/quic/cc_newreno.c

```diff
@@ -0,0 +1,377 @@
+    if (max_init_wnd < MIN_MAX_INIT_WND_SIZE)
+        max_init_wnd = MIN_MAX_INIT_WND_SIZE;
+
+    nr->k_init_wnd = 10 * max_dgram_size;
+    if (nr->k_init_wnd > max_init_wnd)
+        nr->k_init_wnd = max_init_wnd;
+
+    nr->k_min_wnd = 2 * max_dgram_size;
+
+    if (is_reduced)
+        nr->cong_wnd = nr->k_init_wnd;
+}
+
+static void newreno_reset(OSSL_CC_DATA *cc)
+{
+    OSSL_CC_NEWRENO *nr = (OSSL_CC_NEWRENO *)cc;
+
+    nr->k_loss_reduction_factor_num     = 1;
+    nr->k_loss_reduction_factor_den     = 2;
+    nr->persistent_cong_thresh          = 3;
+
+    nr->cong_wnd                    = nr->k_init_wnd;
+    nr->bytes_in_flight             = 0;
+    nr->bytes_acked                 = 0;
+    nr->slow_start_thresh           = UINT64_MAX;
+    nr->cong_recovery_start_time    = ossl_time_zero();
+
+    nr->processing_loss         = 0;
+    nr->tx_time_of_last_loss    = ossl_time_zero();
+    nr->in_congestion_recovery  = 0;
+}
+
+static int newreno_set_option_uint(OSSL_CC_DATA *cc, uint32_t option_id,
+                                   uint64_t value)
+{
+    OSSL_CC_NEWRENO *nr = (OSSL_CC_NEWRENO *)cc;
+
+    switch (option_id) {
+    case OSSL_CC_OPTION_MAX_DGRAM_PAYLOAD_LEN:
+        if (value > SIZE_MAX || value < QUIC_MIN_INITIAL_DGRAM_LEN)
+            return 0;
+
+        newreno_set_max_dgram_size(nr, (size_t)value);
+        return 1;
+
+    default:
+        return 0;
+    }
+}
+
+static int newreno_get_option_uint(OSSL_CC_DATA *cc, uint32_t option_id,
+                                   uint64_t *value)
+{
+    OSSL_CC_NEWRENO *nr = (OSSL_CC_NEWRENO *)cc;
+
+    switch (option_id) {
+    case OSSL_CC_OPTION_MAX_DGRAM_PAYLOAD_LEN:
+        *value = (uint64_t)nr->max_dgram_size;
+        return 1;
+
+    case OSSL_CC_OPTION_CUR_CWND_SIZE:
+        *value = nr->cong_wnd;
+        return 1;
+
+    case OSSL_CC_OPTION_MIN_CWND_SIZE:
+        *value = nr->k_min_wnd;
+        return 1;
+
+    case OSSL_CC_OPTION_CUR_BYTES_IN_FLIGHT:
+        *value = nr->bytes_in_flight;
+        return 1;
+
+    case OSSL_CC_OPTION_CUR_STATE:
+        if (nr->in_congestion_recovery)
+            *value = 'R';
+        else if (nr->cong_wnd < nr->slow_start_thresh)
+            *value = 'S';
+        else
+            *value = 'A';
+        return 1;
+
+    default:
+        return 0;
+    }
+}
+
+static int newreno_in_cong_recovery(OSSL_CC_NEWRENO *nr, OSSL_TIME tx_time)
+{
+    return ossl_time_compare(tx_time, nr->cong_recovery_start_time) <= 0;
+}
+
+static void newreno_cong(OSSL_CC_NEWRENO *nr, OSSL_TIME tx_time)
+{
+    /* No reaction if already in a recovery period. */
+    if (newreno_in_cong_recovery(nr, tx_time))
+        return;
+
+    /* Start a new recovery period. */
+    nr->in_congestion_recovery = 1;
+    nr->cong_recovery_start_time = nr->now_cb(nr->now_cb_arg);
+    nr->slow_start_thresh
+        = (nr->cong_wnd * nr->k_loss_reduction_factor_num)
+          / nr->k_loss_reduction_factor_den;
```

Comment:
I'm not sure about the sizes of these but is it be worth using the `internal/safe_math.h` overflow safe combined multiply/divide here?